File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = Integer.MIN_VALUE;
    int dfs(TreeNode node) {
        if(node == null) return 0;
        int curr = node.val;
        int left = dfs(node.left);
        int right = dfs(node.right);
        int locmax = Math.max(left + curr, curr);
        locmax = Math.max(right + curr, locmax);
        max = Math.max(max, locmax);
        max = Math.max(max, left + curr + right);
        return locmax;
    }
    public int maxPathSum(TreeNode root) {
        dfs(root);
        return max;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    ans = -inf
    def dfs(self, node: Optional[TreeNode]) -> int:
        if not node: return 0
        curr = node.val
        left = self.dfs(node.left)
        right = self.dfs(node.right)
        locmax = max(curr, curr + left)
        locmax = max(locmax, curr + right)
        self.ans = max(locmax, self.ans)
        self.ans = max(self.ans, curr + left + right)
        return locmax
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.dfs(root)
        return self.ans